core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -327,16 +328,37 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             case KOTLIN_COROUTINES:
             case SCALA_FUTURE:
                 final CompletableFuture<?> composedFuture;
+                final AtomicReference<CompletionStage<?>> upstreamStage = new AtomicReference<>();
```

Comment:
Could we remove `AtomicReference` if we create a response future early?
For example:
```java
final CompletableFuture<HttpResponse> responseFuture = new CompletableFuture<>();
...
composedFuture = f.thenCompose(
                            // Add a callback that cancels upstream future in `toCompletionStage(...)`
                            aReq -> toCompletionStage(invoke(ctx, req, aReq), ctx.eventLoop(), responseFuture));

composedFuture.handle((result, cause) -> {
    if (cause != null) {
        responseFuture.completeExceptionally(cause);
    } else {
        responseFuture.complete(convertResponse(ctx, null, result, HttpHeaders.of()));
    }
    return null;
});
```

Reply:
Awesome, looks more reasonable!

Refactored as per the suggestion!
But, I'd decided to introduce a separate method for cancellation propagation logic rather than having it into `toCompletionStage()`.